examples/hashmap.dx

```diff
@@ -0,0 +1,266 @@
+
+
+' # broken because https://github.com/google-research/dex-lang/issues/340
+-- should be a fallback for anything that doesn't have a better definition of its own
+```
+@instance FallbackHashable : Eq t ?=> Hashable t = (MkHashable
+    (\a b. a==b)
+    (\seed x. 1)  -- performance degrades to O(n) search 🤷
+)
+```
+```
+:p thehash (1, 2)
+:p thehash (10, 20)
+```
+
+' ## HashDict
+Now we can build a HashMap on top of that ListDict.
+We are going to hard-code a bunch of buffer sizes for simplicity
+
+data HashDict k:Type v:Type =
+  HashDictBuckets n:Int buckets:(Fin n=>(ListDict k v))
+
+
+def emptyHashDict (n:Int) (k:Type) (v:Type) : HashDict k v =
+    HashDictBuckets n for i:(Fin n). emptyListDict k v
+
+:p emptyHashDict 8 Int Char
+
+def storeHashDict (_:Hashable k) ?=> (dict:HashDict k v) (key:k) (val:v) : HashDict k v =
+    (HashDictBuckets nbuckets buckets) = dict
+    keyhash = thehash key
+    bucket_index = (keyhash `mod` nbuckets)@(Fin nbuckets)
+    bucket = buckets.bucket_index
+    neo_buckets = snd $ withState buckets \bucketsRef.
+        bucketsRef!bucket_index := storeListDict bucket key val
+    HashDictBuckets _ neo_buckets
+
+hd = emptyHashDict 4 Char Int
+
+hd1 = storeHashDict hd 'a' 200
+hd2 = storeHashDict hd1 'b' 201
+hd3 = storeHashDict hd2 'c' 202
+hd4 = storeHashDict hd3 'd' 203
+hd5 = storeHashDict hd4 'e' 204
+
+:p hd5
+
+def tryGetHashDict (_:Hashable k) ?=> (dict:HashDict k v) (key:k) : Maybe v =
+    (HashDictBuckets nbuckets buckets) = dict
+    keyhash = thehash key
+    bucket_index = (keyhash `mod` nbuckets)@(Fin nbuckets)
+    bucket = buckets.bucket_index
+    tryGetListDict bucket key
+
+
+:p tryGetHashDict hd5 'b'
+:p tryGetHashDict hd5 'd'
+
+
+' Benchmarking
+https://github.com/google-research/dex-lang/issues/341
+
+full_data = for i:(Fin 256). ([10 * ordinal i, ordinal i], i)
+
+big_listdict = fold (emptyListDict _ _)  \i state. storeListDict state (fst full_data.i) (snd full_data.i)
+
+
+big_hashdict = fold (emptyHashDict 128 _ _) \i state. storeHashDict state (fst full_data.i) (snd full_data.i)
+
+%time
+tryGetHashDict big_hashdict [1280, 128]
+
+%time
+tryGetListDict big_listdict [1280, 128]
+
+%time
+thehash [1280, 128]
+
+
+' ## Wrap under a common interface
+See how annoying above was without a common interface that i could just pass in the initial value (with its type) and have it work out which store to read etc
+TODO: once this is working remove stand-alone manually type-specied functions from the ListDict and the HashDict
+
+
+interface Associative dict:Type k:Type v:Type where
+    tryGet : dict -> k -> Maybe v
+    store : dict -> k -> v -> dict
+
+
+instance listDictAssociative : Hashable k ?=>  Associative (ListDict k v) k v where
+    store = storeListDict
+    tryGet = tryGetListDict
+
+instance HashDictAssociative : Hashable k ?=>  Associative (HashDict k v) k v where
+    store = storeHashDict
+    tryGet = tryGetHashDict
+
+
+' all of the following error except the first
```

Comment:
This comment is wrong, only the `tryGet` error.
but I don't know why they do

```
:p tryGet eg_listdict 'a'
Type error:Ambiguous type variables: [?2]

( ()
, ( ( tmp @> ( ((k:Type)
                ?-> (v:Type)
                ?-> (Associative (ListDict Word8 Int32) k v)
                ?=> (ListDict Word8 Int32) -> k -> Maybe v)
, (tryGet (ListDict Word8 Int32)) )
, tmp1 @> ( ((v:Type)
             ?-> (Associative (ListDict Word8 Int32) Word8 v)
             ?=> (ListDict Word8 Int32) -> Word8 -> Maybe v)
, (tmp Word8) )
, tmp2 @> ( ((Associative (ListDict Word8 Int32) Word8 ?2)
             ?=> (ListDict Word8 Int32) -> Word8 -> Maybe ?2)
, (tmp1 ?2) )
, tmp3 @> (((ListDict Word8 Int32) -> Word8 -> Maybe ?2), (tmp2 _))
, tmp4 @> ((Word8 -> Maybe ?2), (tmp3 eg_listdict))
, tmp5 @> ((Maybe ?2), (tmp4 'a'))
, tmp6 @> ((Maybe ?2), tmp5)
, _ans_ @> ((Maybe ?2), tmp6) )
, [ tmp:((k:Type)
         ?-> (v:Type)
         ?-> (Associative (ListDict Word8 Int32) k v)
         ?=> (ListDict Word8 Int32) -> k -> Maybe v) =
      tryGet (ListDict Word8 Int32)
  , tmp1:((v:Type)
          ?-> (Associative (ListDict Word8 Int32) Word8 v)
          ?=> (ListDict Word8 Int32) -> Word8 -> Maybe v) = tmp Word8
  , tmp2:((Associative (ListDict Word8 Int32) Word8 ?2)
          ?=> (ListDict Word8 Int32) -> Word8 -> Maybe ?2) = tmp1 ?2
  , tmp3:((ListDict Word8 Int32) -> Word8 -> Maybe ?2) = tmp2 _
  , tmp4:(Word8 -> Maybe ?2) = tmp3 eg_listdict
  , tmp5:(Maybe ?2) = tmp4 'a'
  , tmp6:(Maybe ?2) = tmp5
  , _ans_:(Maybe ?2) = tmp6 ] ) )
  ```
```

Reply:
I'm pretty sure that this is because Dex isn't aware that the `dict` parameter of `Associative` actually implies the values of `k` and `v`. Nothing prevents you from defining weird instances such as `Associative (HashDict k v) Int Float`, so it conservatively assumes that this code is ambiguous.

Reply:
FWIW, the same problem in Haskell is solved using the FunctionalDependencies extension, but we don't have it in Dex, and I'm not sure if we should. Our type class design is still a bit up in the air.